crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -233,15 +233,50 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
         }
     }
 
-    fn balances(
-        &self,
+    fn balances<'a>(
+        &'a self,
         owner: &Address,
-        base_asset_id: &AssetId,
+        start: Option<AssetId>,
+        base_asset_id: &'a AssetId,
         direction: IterDirection,
-    ) -> BoxedIter<'_, StorageResult<(AssetId, TotalBalanceAmount)>> {
+    ) -> BoxedIter<'a, StorageResult<(AssetId, TotalBalanceAmount)>> {
+        match (direction, start) {
+            (IterDirection::Forward, None) => {
+                let base_asset_balance = self.base_asset_balance(base_asset_id, owner);
+                let non_base_asset_balance =
+                    self.non_base_asset_balances(owner, None, direction, base_asset_id);
+                base_asset_balance
+                    .chain(non_base_asset_balance)
+                    .into_boxed()
+            }
+            (IterDirection::Forward, Some(asset_id)) => {
+                let start = (asset_id != *base_asset_id)
```

Comment:
If `asset_id == base_asset_id`, you will skip base asset while we should return it, since it is starting entry=)

Reply:
Ok, I was wondering why it actually works... :)

It turned out that even if we return the bass asset id here, it'll be skipped by the logic inside `query_pagination` [here](https://github.com/FuelLabs/fuel-core/blob/master/crates/fuel-core/src/schema.rs#L170). Internally, it kinda changes the notion of `start` to actually mean `after`.

So now the [updated](https://github.com/FuelLabs/fuel-core/pull/2490/commits/a943e2be2cee62389a14e4cd5d8e0999ffff0cf2) behavior always includes the `start` entry in the output, even though it'll be filtered by the query. While it won't change the returned balances, it is now consistent with other pagination queries. This consistency might be actually important because the query code makes some further calculations about `has_next_page` and other flags.